components/PortalStatistics.lua

```diff
@@ -0,0 +1,1225 @@
+
+local CURRENCY_FORMAT_OPTIONS = {dashIfZero = true, abbreviation = false, formatValue = true}
+local CURRENT_YEAR = tonumber(os.date('%Y'))
+local DATE = os.date('%F')
+local EPOCH_DATE = '1970-01-01'
+local TIMESTAMP = DateExt.readTimestamp(DATE)
+local DEFAULT_ALLOWED_PLACES = Array.map(mw.text.split('1,2,3,1-2,2-3,2-4,3-4', ',', true), String.trim)
+local DEFAULT_ROUND_PRECISION = Info.defaultRoundPrecision or 2
+local LANG = mw.getContentLanguage()
+local SHOWMATCH = 'Showmatch'
+local TIER1 = '1'
+local FIRST = '1'
+local MODES = {'solo', 'team', 'other'}
+local TYPES = {'Online', 'Offline'}
+local GAMES = Array.map(Array.extractValues(Info.games, Table.iter.spairs), function(value)
+	return value.name
+end)
+
+
+local StatisticsPortal = {}
+
+--[[
+Section: Chart Entry Functions
+]] --
+
+
+function StatisticsPortal.gameEarningsChart(args)
+	args = args or {}
+
+	local params = {
+		variable = 'game',
+		processFunction = StatisticsPortal._defaultProcessFunction,
+		catLabel = 'Year',
+		defaultInputs = GAMES,
+	}
+
+	local config = StatisticsPortal._getChartConfig(args, params)
+	local yearSeriesData = StatisticsPortal._cacheModeEarningsData(config)
+	return StatisticsPortal._buildChartData(config, yearSeriesData, config.customLegend, true)
+end
+
+function StatisticsPortal.modeEarningsChart(args)
+	args = args or {}
+
+	local params = {
+		variable = 'opponenttype',
+		processFunction = StatisticsPortal._defaultProcessFunction,
+		catLabel = 'Year',
+		defaultInputs = MODES,
+	}
+
+	local config = StatisticsPortal._getChartConfig(args, params)
+	local yearSeriesData = StatisticsPortal._cacheModeEarningsData(config)
+	return StatisticsPortal._buildChartData(config, yearSeriesData, config.customLegend, true)
+end
+
+
+function StatisticsPortal.topEarningsChart(args)
+	args = args or {}
+	args.limit = tonumber(args.limit) or 10
+
+	local params = {
+		catLabel = Logic.readBool(args.isForTeam) and 'Teams' or 'Players',
+		flipAxes = true,
+	}
+
+	local config = StatisticsPortal._getChartConfig(args, params)
+	local conditions = ConditionTree(BooleanOperator.all)
+		:add{ConditionNode(ColumnName('earnings'), Comparator.gt, 0)}
+	local topEarningsList
+
+	if config.opponentType == Opponent.team then
+		topEarningsList = StatisticsPortal._getTeams(
+			args.limit,
+			conditions:toString(),
+			'sum::earnings desc',
+			'earnings asc')
+	elseif config.opponentType == Opponent.solo then
+		topEarningsList = StatisticsPortal._getPlayers(
+			args.limit,
+			conditions:toString(),
+			'sum::earnings desc',
+			'earnings asc')
+	end
+
+	local yearSeriesData = Array.map(Array.range(config.startYear, CURRENT_YEAR), function(year)
+		return Array.map(Array.reverse(topEarningsList), function(teamData)
+			return teamData.extradata['earningsin'..year] or 0
+		end)
+	end)
+
+	local opponentNames = Array.map(Array.reverse(topEarningsList), function(opponent)
+		return config.opponentType == Opponent.team and opponent.name or opponent.id
```

Comment:
mind those values do not necessarily have to be unique
there exist players that have the same id as other players (same for teams)

Reply:
Would pagename be better then?

Reply:
pagename would be unique 
but not nice for display
so it depends on the usecase

Reply:
This line only sets the data to be used as the category names on the chart axis, so the earnings data would still be calculated correctly. 

I think we keep it this way until we cross the bridge of two players with the same ID appearing on the top earnings chart.